include/ngp_utils/NgpLoopUtils.h

```diff
@@ -0,0 +1,401 @@
+void run_entity_algorithm(
+  Mesh& mesh,
+  const stk::topology::rank_t rank,
+  const stk::mesh::Selector& sel,
+  const AlgFunctor algorithm)
+{
+  using Traits         = NGPMeshTraits<Mesh>;
+  using TeamPolicy     = typename Traits::TeamPolicy;
+  using TeamHandleType = typename Traits::TeamHandleType;
+  using MeshIndex      = typename Traits::MeshIndex;
+
+  const auto& buckets = mesh.get_bucket_ids(rank, sel);
+  auto team_exec = TeamPolicy(buckets.size(), Kokkos::AUTO);
+
+  Kokkos::parallel_for(
+    team_exec, KOKKOS_LAMBDA(const TeamHandleType& team) {
+      auto bktId = buckets.device_get(team.league_rank());
+      auto& bkt = mesh.get_bucket(rank, bktId);
+
+      const size_t bktLen = bkt.size();
+      Kokkos::parallel_for(
+        Kokkos::TeamThreadRange(team, bktLen),
+        [&](const size_t& bktIndex) {
+          MeshIndex meshIdx{&bkt, static_cast<unsigned>(bktIndex)};
+          algorithm(meshIdx);
+        });
+    });
+}
+
+/** Execute the given functor for all edges in a Kokkos parallel loop
+ *
+ *  The functor is called with one argument MeshIndex, a struct containing a
+ *  pointer to the NGP bucket and the index into the bucket array for this
+ *  entity.
+ *
+ *  @param mesh A STK NGP mesh instance
+ *  @param sel  STK mesh selector to choose buckets for looping
+ *  @param algorithm A functor that will be executed for each entity
+ */
+template<typename Mesh, typename AlgFunctor>
+void run_edge_algorithm(
+  Mesh& mesh,
+  const stk::mesh::Selector& sel,
+  const AlgFunctor algorithm)
+{
+  static constexpr stk::topology::rank_t rank = stk::topology::EDGE_RANK;
+  using Traits    = NGPMeshTraits<Mesh>;
+  using MeshIndex = typename Traits::MeshIndex;
+
+  run_entity_algorithm(
+    mesh, rank, sel,
+    KOKKOS_LAMBDA(MeshIndex& meshIdx) {
+      algorithm(
+        EntityInfo<Mesh>{meshIdx, (*meshIdx.bucket)[meshIdx.bucketOrd],
+            mesh.get_nodes(meshIdx)});
+  });
+}
+
+/** Execute the given functor for all elements in a Kokkos parallel loop
+ *
+ *  The functor is called with one argument MeshIndex, a struct containing a
+ *  pointer to the NGP bucket and the index into the bucket array for this
+ *  entity.
+ *
+ *  Note that a rank is still passed as an argument to allow looping over both
+ *  element and side/face ranks with the same function.
+ *
+ *  @param mesh A STK NGP mesh instance
+ *  @param sel  STK mesh selector to choose buckets for looping
+ *  @param algorithm A functor that will be executed for each entity
+ */
+template<typename Mesh, typename AlgFunctor>
+void run_elem_algorithm(
+  Mesh& mesh,
+  const stk::topology::rank_t rank,
+  const stk::mesh::Selector& sel,
+  const AlgFunctor algorithm)
+{
+  using Traits    = NGPMeshTraits<Mesh>;
+  using MeshIndex = typename Traits::MeshIndex;
+
+  run_entity_algorithm(
+    mesh, rank, sel,
+    KOKKOS_LAMBDA(MeshIndex& meshIdx) {
+      algorithm(
+        EntityInfo<Mesh>{meshIdx, (*meshIdx.bucket)[meshIdx.bucketOrd],
```

Comment:
I agree that MeshIndex is the right thing to use here. The stk team has a task (which we haven't gotten to yet) to clean up the mesh-indexes. We have at least 3 of them. The one you're using here has a bucket-pointer, which means it can't be copied from host to device because the pointer would be bad, but if it's created and used within the device, like you're doing here, then I think it is the best one to use.